types/params.go

```diff
@@ -161,6 +161,26 @@ func DefaultTimeoutParams() TimeoutParams {
 	}
 }
 
+// ProposeTimeout returns the amount of time to wait for a proposal.
+func (t TimeoutParams) ProposeTimeout(round int32) time.Duration {
+	return time.Duration(
+		t.Propose.Nanoseconds()+t.ProposeDelta.Nanoseconds()*int64(round),
+	) * time.Nanosecond
+}
+
+// VoteTimeout returns the amount of time to wait for remaining votes after receiving any +2/3 votes.
+func (t TimeoutParams) VoteTimeout(round int32) time.Duration {
+	return time.Duration(
+		t.Vote.Nanoseconds()+t.VoteDelta.Nanoseconds()*int64(round),
+	) * time.Nanosecond
+}
+
+// CommitTime returns the amount of time to wait for remaining votes after receiving +2/3 precommits
```

Comment:
This returns a `time.Time` (i.e., a moment) but the comment says "amount of time" implying a duration. What I _think_ we mean here might be:

```go
// CommitTime returns the earliest time at which consensus should
// commit the current block. The resulting time will be at or after
// ti, and reflects the interval after receiving +2/3 precommits
// during which the node should wait for additional votes before
// committing the block.
```

or words to that effect. This is tricky enough to explain that I wonder if we shouldn't just make it an interval instead.

Reply:
_Sigh_ This is somewhat a symptom of the kind of awkward way we handle this timeout in the consensus code. We calculate the `StartTime` for the next height, and various methods check if they are before it. Once the consensus code is notified that there are Tx for a block, it schedules a timeout using this `StartTime` value.

https://github.com/tendermint/tendermint/blob/ab0ff9b4aa0f731a7fb877c354ef245b5f14a8cc/internal/consensus/state.go#L790

https://github.com/tendermint/tendermint/blob/ab0ff9b4aa0f731a7fb877c354ef245b5f14a8cc/internal/consensus/state.go#L1118

The other timeouts are a bit different. When, say, the Vote timeout needs to be calculated, it is because the code is directly about to schedule the timeout.

https://github.com/tendermint/tendermint/blob/ab0ff9b4aa0f731a7fb877c354ef245b5f14a8cc/internal/consensus/state.go#L1623

Reply:
Makes sense. In that case it's probably reasonable to leave it as a time for now. The important thing, I think, is to make the comment and the code line up (even if that's somewhat tricky to get right 🙂)

Reply:
Very true, updated. Thanks for the catch 🙏